domains/client/relayer/src/lib.rs

```diff
@@ -80,102 +87,160 @@ impl From<sp_api::ApiError> for Error {
+            );
         }
+    }
 
-        Ok(())
+    Ok(())
+}
+
+/// Sends an Outbox message from src_domain to dst_domain.
+fn gossip_outbox_message<Block, Client, CHash>(
+    client: &Arc<Client>,
+    msg: CrossDomainMessage<CHash, H256>,
+    sink: &GossipMessageSink,
+) -> Result<(), Error>
+where
+    Block: BlockT,
+    CHash: Codec,
+    Client: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+    Client::Api: RelayerApi<Block, NumberFor<Block>, CHash>,
+{
+    let best_hash = client.info().best_hash;
+    let dst_chain_id = msg.dst_chain_id;
+    let ext = client
+        .runtime_api()
+        .outbox_message_unsigned(best_hash, msg)?
+        .ok_or(Error::FailedToConstructExtrinsic)?;
+
+    sink.unbounded_send(GossipMessage {
+        chain_id: dst_chain_id,
+        encoded_data: ext.encode(),
+    })
+    .map_err(Error::UnableToSubmitCrossDomainMessage)
+}
+
+/// Sends an Inbox message response from src_domain to dst_domain
+/// Inbox message was earlier sent by dst_domain to src_domain and
+/// this message is the response of the Inbox message execution.
+fn gossip_inbox_message_response<Block, Client, CHash>(
+    client: &Arc<Client>,
+    msg: CrossDomainMessage<CHash, H256>,
+    sink: &GossipMessageSink,
+) -> Result<(), Error>
+where
+    Block: BlockT,
+    CHash: Codec,
+    Client: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+    Client::Api: RelayerApi<Block, NumberFor<Block>, CHash>,
+{
+    let best_hash = client.info().best_hash;
+    let dst_chain_id = msg.dst_chain_id;
+    let ext = client
+        .runtime_api()
+        .inbox_response_message_unsigned(best_hash, msg)?
+        .ok_or(Error::FailedToConstructExtrinsic)?;
+
+    sink.unbounded_send(GossipMessage {
+        chain_id: dst_chain_id,
+        encoded_data: ext.encode(),
+    })
+    .map_err(Error::UnableToSubmitCrossDomainMessage)
+}
+
+impl<Client, Block> Relayer<Client, Block>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block> + AuxStore + ProofProvider<Block> + ProvideRuntimeApi<Block>,
+{
+    /// Constructs the proof for the given key using the consensus chain backend.
+    fn construct_consensus_chain_xdm_proof_for_key_at(
+        consensus_chain_client: &Arc<Client>,
+        finalized_block: (NumberFor<Block>, Block::Hash),
+        block_hash_to_process: Block::Hash,
+        key: &[u8],
+    ) -> Result<ProofOf<Block>, Error>
+    where
+        Client::Api: MmrApi<Block, H256, NumberFor<Block>>,
+    {
+        let proof = consensus_chain_client
+            .read_proof(block_hash_to_process, &mut [key].into_iter())
+            .map_err(|_| Error::ConstructStorageProof)?;
+
+        let (mmr_leaf, mmr_proof) =
+            construct_consensus_mmr_proof(consensus_chain_client, finalized_block.0)?;
+
+        Ok(Proof::Consensus {
+            consensus_chain_mmr_proof: ConsensusChainMmrLeafProof {
+                consensus_block_hash: finalized_block.1,
```

Comment:
This is unused because the MMR proof is self contained but I wanted to keep the hash of finalized block for light client or other who are tracking MMR roots to be able to verify the proof stateless